File: JAVA/14_Backtracking/BOJ_1759_암호만들기.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;
import java.util.Arrays;

public class Main {

    static int l, c;
    static char[] arr, password;
    static boolean[] visited;
    static StringBuilder sb;

    public static void dfs(int dpth) {
        if (dpth == l) {
            StringBuilder sb2 = new StringBuilder();
            int vowel = 0, consonant = 0;
            for (int i = 0; i < l; i++) {
                sb2.append(arr[i]);
                if (arr[i] == 'a' || arr[i] == 'e' || arr[i] == 'i' || arr[i] == 'o' || arr[i] == 'u') {
                    vowel++;
                } else consonant++;
            }
            if (vowel > 0 && consonant >= 2) {
                sb.append(sb2).append("\n");
            }
            return;
        }
        for (int i = 0; i < c; i++) {
            if (!visited[i]) {
                if (dpth > 0 && arr[dpth - 1] > password[i]) continue;
                visited[i] = true;
                arr[dpth] = password[i];
                dfs(dpth + 1);
                visited[i] = false;
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        l = Integer.parseInt(st.nextToken());
        c = Integer.parseInt(st.nextToken());
        arr = new char[l];
        password = new char[c];
        visited = new boolean[c];
        st = new StringTokenizer(br.readLine(), " ");
        for (int i = 0; i < c; i++) {
            password[i] = st.nextToken().charAt(0);
        }
        Arrays.sort(password);
        sb = new StringBuilder();
        dfs(0);
        System.out.print(sb);
    }
}

File: C/14_Backtracking/BOJ_1759_암호만들기.c
#include <stdio.h>
int visited[15], l, c;
char arr[15], password[15];
void quickSort(char arr[], int left, int right) {
	int i = left, j = right, pivot = arr[(left + right) / 2];
	char temp;
	do {
		while (arr[i] < pivot) i++;
		while (arr[j] > pivot) j--;
		if (i <= j) {
			temp = arr[i];
			arr[i] = arr[j];
			arr[j] = temp;
			i++;
			j--;
		}
	} while (i <= j);
	if (left < j) quickSort(arr, left, j);
	if (i < right) quickSort(arr, i, right);
}
void dfs(int dpth) {
	if (dpth == l) {
		int vowel = 0, consonant = 0;
		for (int i = 0; i < l; i++) {
			if (arr[i] == 'a' || arr[i] == 'e' || arr[i] == 'i' || arr[i] == 'o' || arr[i] == 'u') {
				vowel++;
			}
			else consonant++;
		}
		if (vowel > 0 && consonant >= 2) {
			for (int i = 0; i < l; i++) {
				printf("%c", arr[i]);
			}
			printf("\n");
		}
		return;
	}
	for (int i = 0; i < c; i++) {
		if (visited[i] == 0) {
			if (dpth > 0 && arr[dpth - 1] > password[i]) continue;
			visited[i] = 1;
			arr[dpth] = password[i];
			dfs(dpth + 1);
			visited[i] = 0;
		}
	}
}
main() {
	scanf("%d %d", &l, &c);
	for (int i = 0; i < c; i++) {
		getchar();
		scanf("%c", &password[i]);
	}
	quickSort(password, 0, c - 1);
	dfs(0);
}